source/auth/Credentials.cpp

```diff
@@ -106,7 +116,7 @@ namespace Aws
 
             CredentialsProvider::~CredentialsProvider()
             {
-                if (m_provider != nullptr)
+                if (m_provider)
```

Comment:
This is mainly to raise the question:
Do we need to call `aws_credentials_provider_acquire(provider)` as well in the constructor above?
To me it seems we should. If we should, there are several places in this repo that needs to be cleaned up, including operations on `aws_credentials` ,`aws_credentials_provider`, etc.

Reply:
_new() always returns an object with a ref count of 1, so there's no need to call acquire on it.

Reply:
I don't see anything relevant to aws_credentials but I suppose you could rework the provider construction to acquire in s_CreateWrappedProvider and release in all of the helper functions.  Anyone else using the public constructor is going to be broken though.

Reply:
Yes, that's what I encountered in the test when I tried to rework this.
What I mean is in this ctor
```
CredentialsProvider::CredentialsProvider(aws_credentials_provider *provider, Allocator *allocator) noexcept
  : m_allocator(allocator), m_provider(provider)
{	            
}	            
```
If we provide customer this ctor, without increasing ref_cnt to the raw provider. It likely will hold an invalid pointer later. There are similar cases for aws_credentials at other places. I didn't list here.

Reply:
Could you point out the similar cases because I didn't see any.

Reply:
After spotting the `aws_credentials_provider` issue, I did a scan of `GetUnderlyingHandle()` call and thought the following code is problematic. Now I realize that I was wrong, m_credentials is a shared_ptr, as long as it stays, m_config.credentials won't be destroyed.
```
void AwsSigningConfig::SetCredentials(const std::shared_ptr<Credentials> &credentials) noexcept
{
        m_credentials = credentials;
        m_config.credentials = m_credentials->GetUnderlyingHandle();
}
```